File: python/plates_between_candles.py
class plater_between_candles :
    def platesBetweenCandles(self, s: str, queries: List[List[int]]) -> List[int]:
        n = len(s)
        
        # Prefix sum for plates
        prefix = [0] * (n + 1)
        for i, ch in enumerate(s):
            prefix[i + 1] = prefix[i] + (1 if ch == '*' else 0)
        
        # Nearest candle to the left
        left_candle = [-1] * n
        last_candle = -1
        for i, ch in enumerate(s):
            if ch == '|':
                last_candle = i
            left_candle[i] = last_candle
        
        # Nearest candle to the right
        right_candle = [-1] * n
        next_candle = -1
        for i in range(n - 1, -1, -1):
            if s[i] == '|':
                next_candle = i
            right_candle[i] = next_candle
        
        answer = []
        for L, R in queries:
            left = right_candle[L]
            right = left_candle[R]
            
            if left != -1 and right != -1 and left < right:
                answer.append(prefix[right + 1] - prefix[left + 1])
            else:
                answer.append(0)
        
        return answer

File: java/platesBetweenCandles.java
package java;

public class platesBetweenCandles {
     public int[] platesBetweenCandle(String s, int[][] queries) {
         int n = s.length();
        
        // Precompute arrays
        int[] prefix = new int[n + 1];
        int[] leftCandle = new int[n];
        int[] rightCandle = new int[n];
        
        // Single forward pass for prefix and left candle
        int lastCandle = -1;
        for (int i = 0; i < n; i++) {
            prefix[i + 1] = prefix[i] + (s.charAt(i) == '*' ? 1 : 0);
            if (s.charAt(i) == '|') lastCandle = i;
            leftCandle[i] = lastCandle;
        }
        
        // Reverse pass for right candle
        int nextCandle = -1;
        for (int i = n - 1; i >= 0; i--) {
            if (s.charAt(i) == '|') nextCandle = i;
            rightCandle[i] = nextCandle;
        }
        
        // Process queries with direct array access
        int[] answer = new int[queries.length];
        for (int i = 0; i < queries.length; i++) {
            int[] q = queries[i];
            int left = rightCandle[q[0]];
            int right = leftCandle[q[1]];
            
            if (left != -1 && right != -1 && left < right) {
                answer[i] = prefix[right] - prefix[left + 1];
            }
        }
        return answer;
    }
}
